Replaced PushStage based Drop with GraphStage #19834

trautonen opened the pull request:
Replaced `PushStage` based `Drop` with a `GraphStage` implementation using `SimpleLinearGraphStage` as a base.

Commit:
Replaced PushStage based Drop with GraphStage #19834

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
```diff
@@ -142,14 +142,27 @@ private[akka] final case class Take[T](count: Long) extends PushPullStage[T, T]
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    private var left: Long = count
+
+    setHandler(in, new InHandler {
```
You can save two allocations here by implementing the `InHandler` and `OutHandler` with in the `GraphStageLogic` itself, and use that `setHandlers(in, out, this)`

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterStressSpec.scala:
```diff
@@ -80,7 +80,7 @@ class InterpreterStressSpec extends AkkaSpec with GraphInterpreterSpecKit {
     }
 
-    "work with a massive chain of drops" in new OneBoundedSetup[Int](Vector.fill(chainLength / 1000)(Drop(1))) {
+    "work with a massive chain of drops" in new OneBoundedSetup[Int](Vector.fill(chainLength / 1000)(Drop(1)): _*) {
```
Since instances of `GraphStageLogic` are reusable you can make one instance of `Drop` and re-use it 1000 times here.

drewhk replied:
Yes, this is a difference from PushPullStage (which required an explicit external factory)

rkuhn replied:
we might be so insidious to cache or even pre-create a few (like `Drop(1)`)

johanandren commented:
OK TO TEST

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
```diff
@@ -142,14 +142,27 @@ private[akka] final case class Take[T](count: Long) extends PushPullStage[T, T]
+        if (left > 0) {
+          left -= 1
+          pull(in)
+        } else push(out, grab(in))
```
You can do `passAlong(in, out)` which would set the two ports in pass-through mode, removing the branch.

rkuhn replied:
that does an allocation, though, so I wouldn’t want to guess which one is preferable without micro-benchmark

rkuhn commented:
LGTM after folding the two handlers into the logic.

trautonen commented:
Updated the pull request with suggested changes:
- Implemented `InHandler` and `OutHandler` in `GraphStageLogic`
- Reused same `Drop` stage instance for stress test

johanandren commented:
LGTM, thanks @trautonen!

johanandren commented:
Refs #19834
